code/modules/projectiles/gun_system.dm

```diff
@@ -181,9 +194,34 @@
 
 /obj/item/weapon/gun/equipped(mob/user, slot)
 	unwield(user)
-
+	if(ishandslot(slot))
+		gun_user = user
+		RegisterSignal(gun_user, COMSIG_MOB_MOUSEDOWN, .proc/start_fire)
+		RegisterSignal(gun_user, COMSIG_MOB_MOUSEUP, .proc/stop_fire)
+		RegisterSignal(gun_user, COMSIG_MOB_MOUSEDRAG, .proc/change_target)
+		RegisterSignal(src, COMSIG_GUN_MUST_FIRE, .proc/Fire)
+		RegisterSignal(src, COMSIG_GUN_FIRE_RESET, .proc/reset_fire)
+	else
+		if(gun_user)
+			UnregisterSignal(gun_user, COMSIG_MOB_MOUSEDOWN, .proc/start_fire)
+			UnregisterSignal(gun_user, COMSIG_MOB_MOUSEUP, .proc/stop_fire)
+			UnregisterSignal(gun_user, COMSIG_MOB_MOUSEDRAG, .proc/change_target)
+			UnregisterSignal(src, COMSIG_GUN_MUST_FIRE, .proc/Fire)
+			UnregisterSignal(src, COMSIG_GUN_FIRE_RESET, .proc/reset_fire)
+			gun_user = null
+		SEND_SIGNAL(src, COMSIG_GUN_AUTO_FIRE_COMP_RESET)
 	return ..()
 
+/obj/item/weapon/gun/removed_from_inventory(mob/user)
+	if(gun_user)
+		UnregisterSignal(gun_user, COMSIG_MOB_MOUSEDOWN, .proc/start_fire)
+		UnregisterSignal(gun_user, COMSIG_MOB_MOUSEUP, .proc/stop_fire)
+		UnregisterSignal(gun_user, COMSIG_MOB_MOUSEDRAG, .proc/change_target)
+		UnregisterSignal(src, COMSIG_GUN_MUST_FIRE, .proc/Fire)
+		UnregisterSignal(src, COMSIG_GUN_FIRE_RESET, .proc/reset_fire)
+		gun_user = null
+	SEND_SIGNAL(src, COMSIG_GUN_AUTO_FIRE_COMP_RESET)
+
```

Comment:
Signals should send early + early returns + please check if this proc is called on mob deletion so gunuser doesnt cause hard dels